### PR TITLE
Comics: zoom to panel

### PR DESCRIPTION
**Pull request by Galunid**
This pull requests aims to provide convenient way to zoom in comics. The idea is when user holds/double taps (not decided yet) on a manga/comic panel, it gets cut out from the rest of the image and zoomed. More details in https://github.com/koreader/koreader-base/issues/1148. Depends on https://github.com/koreader/koreader-base/pull/1159

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6511)

### CODE REVIEW

**Comment by poire-z** [comment]
I won't be able to comment much on the kopt/leptonica stuff as I'm not familiar with it.
Just curious: for a page with 3 full-width panels, stacked vertically, what will this do ? Will there be a zoom ? Or just one of the 3 panels extracted and shown standalone not zoomed as it takes the screen width ?

**Comment by Galunid** [comment]
Right now, it depends on the specific case. It will probably detect one page as one panel, and "panel of that page" as another, so it depends which one is "earlier" in the array. Additionally there is a check whether the panel is bigger than 1/10 of width/height, so it's possible the smaller one will get rejected. So far, this only triggers image detection in `KOPTContext` and it writes to files with index numbers. I'm figuring out how to get image/boundary from `KOPTContext` back to ReaderHighlight, so it can be displayed. It shouldn't be difficult, but I'm taking a break from programming/studying for today ;-)

**Comment by poire-z** [comment]
(While browsing old issues, I just met #2478 - so just mentionning it here as it's also about detecting panels.)

**Comment by Galunid** [comment]
Assuming they are within "borders" (black rectangles enclosing the text), it will work on that too! ;-)

**Comment by Frenzie** [comment]
I think those borders are merely illustrative. :-)

**Comment by Galunid** [comment]
Welp, there are some examples of page segmentation in leptonica, so I may take a look after I'm done with this

**Comment by Galunid** [comment]
Is there a way to set default value format-wise? What I mean is for `.cbz` to have different value than `.epub/.pdf`.
If anyone is interesting in testing it needs https://github.com/koreader/koreader-base/pull/1159 and disabling highlighting. Any ideas on how to add this to menu greatly appreciated.

**Comment by Frenzie** [comment]
> Is there a way to set default value format-wise? What I mean is for .cbz to have different value than .epub/.pdf.

You mean currently? I don't think so; we pretty much do things engine-wise, and CBZ is handled by MuPDF. I've thought about it a little bit (in the context of MuPDF's EPUB support) but never implemented anything.

**Comment by ezdiy** [comment]
Tangentially related: Reading comic books on 7" is not always all that pleasant not necessarily because I'd care enough to zoom into panels, but because the bubble font is too small. So I have this idea floating around to:

1) Identify text bubbles
2) Auto-detect if font is too small (and hallucinate zoom factor from it) in there best scenario, but good enough would be even just user setting per book
2) Zoom text in there just by stretching it (ocr too slow/inaccurate). This should look ok because the pages are usually far higher resolution than native device one anyway. Still, needs to "organically" enlarge the bubble area if need be to fit the zoomed text.

Would leptonica be up to the task, or does this mean bringing more image processing in?

**Comment by Galunid** [comment]
Leptonica should be able to do that. The algorithm I have in mind would be something like detect text regions, remove them, use edge detection algorithm -> crop and zoom. We could do it even faster, assuming user "holds the speech bubble". In that case we only need to find text only in that region. I'd like to finish implementing EDP controller for Onyx devices, since whether I keep it or not depends on that and I have 2 days left to decide. After that I'll finish this PR and try to add some documentation to stuff related to Leptonica. I can try to implement your suggestion when I'm done with that. It would be cool to have double tap to zoom to panel, hold to zoom to speech bubble.

**Comment by Frenzie** [comment]
> We could do it even faster, assuming user "holds the speech bubble".

Potentially sounds like overcomplicating things compared to just zoom to region. ^_^

**Comment by ezdiy** [comment]
Yep, the problem is that whenever you'd want to do this, you'd end up tapping all bubbles anyway. I'd prefer to limit interaction with the page to no interaction (as my device input speed/accuracy is ... not great). Hence why the wholesale "for each bubble, detect bubble clipping mask, zoom that, plaster it over center of the original bubble). The point is to deal scanlations where typesetting isn't particularly great.

**Comment by Galunid** [comment]
Makes sense, I'll see what I can do.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 1]
```diff
@@ -595,13 +595,21 @@ function ReaderHighlight:onPanelZoom(arg, ges)
...
     end
+    logger.dbg("File:", self.ui.document.file)
+    logger.dbg("Extension:", util.getFileNameSuffix(self.ui.document.file))
     return true
 end
 
 function ReaderHighlight:onHold(arg, ges)
+    -- if it's manga/comic, panel zoom
+    local filetype = util.getFileNameSuffix(self.ui.document.file)
+    if filetype == 'cbz' or filetype == 'djvu' then
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:606]: Nit:
```suggestion
    if filetype == "cbz" or filetype == "djvu" then
```

This is just a proof of concept atm I assume? Otherwise I might have bigger comments. :-P

Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua]: It's PoC for now, I'll try to get it ready for review tomorrow, when I add menu entry to enable/disable this feature (and figure out where to put it ;-))

**Comment by Galunid** [comment]
I put the "toggle" option in `Highlighting` menu, I couldn't really think of a better place, but suggestions are really welcome here, should be ready for review.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 2]
```diff
@@ -596,10 +611,13 @@ function ReaderHighlight:onPanelZoom(arg, ges)
     return true
 end
 
+local function isDocumentComicOrManga(file)
+    local filetype = util.getFileNameSuffix(file)
+    return filetype == "cbz" or filetype == "djvu"
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:616]: I can see treating CBZ as a comic by default, but DjVu is basically the same as PDF here? I read most of my manga/comics in PDF. On a related note, we also support CBT.

Comment (ezdiy) [reply frontend/apps/reader/modules/readerhighlight.lua:616]: `filetype:match("^cb[zrt7]") or filetype=="zip"`

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerhighlight.lua:616]: We don't currently support CBR and CB7(or do we? :D), and zip is potentially a bit of an issue for zipped FB2 or something?

But, yeah, +1 on the cb pattern regardless ;).

Comment (ezdiy) [reply frontend/apps/reader/modules/readerhighlight.lua]: @NiLuJe mupdf can do it, provided it's built that way. As for zip, it's extremely common format for ... "certain" kinds of manga :>

That said whether it's zip or fb2 inside is discerned by mupdf again, so perhaps more reliable method would be to inquire mupdf about what's it doing.

Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua]: I think for now, it's better not to support `djvu`, since they sometimes fail, when `cbz` works (due to lower resolution believe). I just tested `cbt` and it works without problems. I'm not sure about `cb7`, but `cbr` isn't supported, right?

Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua]: @NiLuJe zipped fb2 don't have extension of `.fb2.zip`?

Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua]: Perhaps a better solution would be to add a separate menu entry, where we let user decide which formats they want supported?

**Comment by NiLuJe** [comment]
I like that idea, yeah ;).

**Comment by Galunid** [comment]
Red rectangle is what's enabled by default, is that ok?
image

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 2 above)
Comment (Frenzie) [reply frontend/apps/reader/modules/readerhighlight.lua]: I think this whole notion of support is weird when it seems like a feature similar to reflow. We talked about defaults earlier.

Comment (Frenzie) [reply frontend/apps/reader/modules/readerhighlight.lua]: > Perhaps a better solution would be to add a separate menu entry, where we let user decide which formats they want supported?

Strange, GH didn't show this to me. Anyway, I'd say it should be a per document thing (cf. reflow), possibly with a per format default.

**Comment by Frenzie** [comment]
Turn on panel zoom

**Comment by poire-z** [comment]
> I think this whole notion of support is weird when it seems like a feature similar to reflow. We talked about defaults earlier.

I kind of agree, it doesn't feel right to tie features to extensions, even if we let the user choose the extensions.
I don't read comics, so I haven't followed much this PR. But tell me if I get it right:
when long press on a panel (the 2-6 small images in a comic page), it should zoom that vignette?
Nothing to do with the text bubbles, right?
So, can't we have both working at the same time?:
- if long press at some position, and text is found at that position: do normal text selection action (dict lookup if single word, otherwise show dialog)
- if no text dedected at position, and if panel zoom is enabled, do panel detection and zoom

(Dunno if we can detect there's text at position, or if the stuff will look for nearest text...)

**Comment by Frenzie** [comment]
@poire-z My PDF Humble Bundle comics often have selectable text. (But I've only read a manga on my device once, Cells at Work. Most stuff isn't b&w so unsuitable.) I prefer the PDFs because they render sharper, with vector text among other things.

But also atm I think it'll just select the nearest text, even if it's pretty far away?

**Comment by poire-z** [comment]
> But also atm I think it'll just select the nearest text, even if it's pretty far away?

So, may be when "panel detection" is enabled, it should additionally to not go too far away :) 
I don't know if that's possible, just throwing some ideas, @Galunid might now more what's possible :)

Btw, be sure that new menu item "Panel zoom" only shows with `document.info.has_pages`, or does it work on EPUBs?

**Comment by ezdiy** [comment]
> (Dunno if we can detect there's text at position, or if the stuff will look for nearest text...)

koreader doesn't have a robust scene text detection. That's a thing you throw whole page at and it tells you *all* the rect that most likely are text. Algorithms for this exist and are resonably simlpe (no need for fancy NNs), but not sure if tesseract alone expose any.

**Comment by Galunid** [comment]
@poire-z I had little luck selecting text in comics, more often than not, it selects part of the image


> when long press on a panel (the 2-6 small images in a comic page), it should zoom that vignette?
> Nothing to do with the text bubbles, right?

Yes, it has nothing to do with text bubbles


> So, may be when "panel detection" is enabled, it should additionally not go too far away :)
> I don't know if that's possible, just throwing some ideas, @Galunid might know more what's possible :)

That's how it's supposed to work. The only issue is when 2 panels are "connected" (by text bubble for example), then it'll select 2 panels. Like this
image
For now I've read some papers on how to deal with that. [This one](http://visal.cs.cityu.edu.hk/static/pubs/conf/mm14-panels.pdf) seems promising, but I haven't gotten around implementing anything. 

@Frenzie 
>  Anyway, I'd say it should be a per document thing (cf. reflow), possibly with a per format default.

I kind of disagree, usually when someone reads manga series they binge it I think (at least I do), so it'd be pretty comfortable to say enable this for all pdf's and disable it once I'm done. I assume `per document` would mean I have to enable it for every single file. Since series are often 100+ files, I don't find the idea appealing.

**Comment by Galunid** [comment]
> I kind of agree, it doesn't feel right to tie features to extensions, even if we let the user choose the extensions.

Would you mind elaborating a bit? I don't really see the downside.



> Btw, be sure that new menu item "Panel zoom" only shows with `document.info.has_pages`, or does it work on EPUBs?

Doesn't work in epub, I forgot to add the condition for displaying it. I will do that soon.

**Comment by poire-z** [comment]
> Would you mind elaborating a bit? I don't really see the downside.

Just that we didn't do anywhere that kind of features/extension tie before this PR (at most, we do it for picdocument/is_pic, that we assume are non-text images).
I see it's practical, so one can enable it for cbz while still reading PDF.
But in the absolute, a cbz may contain text images, and a pdf may be a comic, and the user would have to tweak these settings when meeting these non-classic cbz and pdf.
Ideally, you would detect this per page, or when needed by finger position.

**Comment by Frenzie** [comment]
> I kind of disagree, usually when someone reads manga series they binge it I think (at least I do), so it'd be pretty comfortable to say enable this for all pdf's and disable it once I'm done. I assume per document would mean I have to enable it for every single file. Since series are often 100+ files, I don't find the idea appealing.

No, per document means the ability to override whatever you set as the default. ;-) Any per-document setting should have a long press set default setting associated with it.

**Comment by Galunid** [comment]
@Frenzie So you're not against associating those settings with extensions, but against a way in which they are presented? Do you just want me to move it from top menu to the bottom menu near reflow with on/off switch?

**Comment by Frenzie** [comment]
No, there have to be per-document settings. That's unrelated to what you're talking about.

* Normal means per-document and global default.
* Custom would mean per-document and extension-specific default.

**Comment by ezdiy** [comment]
+1 per document, just like stroke boldness font size etc. All of these are tracked by default per document, but if you long-press you configure the default value for documents that don't set their own. This is far more ergonomic than any settings menu.

**Comment by Galunid** [comment]
Do I get it right? There is global setting with a list of extensions where `panel zoom` is enabled by default. Then there is a document wide setting that allows for toggling `panel zoom` on/off for that specific document. When I hold on that document-wide setting button, it should change global setting?

**Comment by poire-z** [comment]
I think you should get rid of the extension selection, and just have an independant menu item with toggle checkbox (i.e. like Floating punctuation):
- tap on that menu item : save state in docsetting (per-document)
- hold on that menu item : save state in settings.reader.lue (global)
- open a new document: use the default value from settings.reader.lua, or your default value (and may be, if there's no default value, start with true for CB* and false for anything else) (and the thing to discuss: whether to save it in docsettings or not)

So, when you're in a Comics reading campaign, you just hold on that menu to set it globally to true. If in this campaign, you open a text PDF, you sut tap on that menu to change it for this book. Open a new CBZ and it opens with the global value.
When you're done with comics, and plan reading more text, you just hold to set it globally to false.

It would feel more accessible in the bottom menu - but the bottom menu is essentially about rendering options... So, having it here fell strange, and I can't find any bottom panel where it would logically fit.

**Comment by Frenzie** [comment]
It's in roughly the same domain as something like crop or reflow, although it's not exactly about rendering. But also I think it would just feel nicely accessible there. :-)

tl;dr What @poire-z said.

**Comment by poire-z** [comment]
And don't forget we have gestures, so if you make it a dispatcher action, toggling between text selection and panel zoom is just a gesture away.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 3]
```diff
@@ -125,8 +126,12 @@ local function isDocumentComicOrManga(file)
     return supported_filetypes[filetype]
 end
 
-local function isPanelZoomEnabled()
-    return G_reader_settings:nilOrTrue("panel_zoom_allowed")
+local function isPanelZoomEnabled(file)
+    local doc_setting = DocSettings:readSetting("panel_zoom_enabled")
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:130]: You can access directly the setting for the currently opened book with:
`self.ui.doc_settings:readSetting("panel_zoom_enabled")`
without re-opening it (and it it has been updated but not yet flushed, you get the updated value).
But as you probably will set the menu to toggle it in this file, you'll probably keep a copy as self.panel_zoom_enabled that you can access directly - and load/save it in the existing:
```lua
function ReaderHighlight:onReadSettings(config)
function ReaderHighlight:onSaveSettings()
```

Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua]: Thanks, saved me quite a bit of time!

**Comment by Galunid** [comment]
Anyone knows how to add stuff to the bottom menu? I was checking and it seems `frontend/ui/data/koptoptions.lua` is responsible for that, but it works differently(?) from the top menu and I don't really see a simple way to include things from `ReaderHighlight` there.

**Comment by poire-z** [comment]
> I don't really see a simple way to include things from ReaderHighlight there.

Yes, we don't populate the bottom menu, like we do for the top menu.
The bottom panels are standalone and defined in koptoptions.lua / creoptions.lua. In there, you tell everything about what you want in the bottom panels, and you tell which events are fired. Then, in ReaderHighlight, you just add the `onEventName()` method that will trigger when that event is fired.
(There's a bit of confusion about which from the bottom pannel and the target module is responsible for saving/loading the state of the toggle - the bottom panel is - and there's also an intermediate readerkoptlistener - so you'll see sometimes that a setting is saved twice with a different name :) it's a bit of a mess.)

For the PDF bottom panel, I don't know where you plan to add your toggle if you really want it there (again, mostly rendering stuff). Try to keep the height of each panel small: the max is 4 widgets par panel currently, you could go to 5 if needed, but if you can fill the 2 left panels that have 1 or 2 items, that would be nicer - althought panel zooming is a bit unrelated, except for the bottom panels icons :) the 2 left ones could visually evoke a comic panel extraction/cropping:
image :)

**Comment by NiLuJe** [comment]
I don't recall what's currently in it, but, yeah, that'd certainly make sense in the crop tab ;).

File: frontend/dispatcher.lua [hunk 4]
```diff
@@ -103,7 +103,8 @@ local settingsList = {
     cycle_highlight_style = { category="none", event="CycleHighlightStyle", title=_("Cycle highlight style"), rolling=true, paging=true, separator=true,},
     kosync_push_progress = { category="none", event="KOSyncPushProgress", title=_("Push progress from this device"), rolling=true, paging=true,},
     kosync_pull_progress = { category="none", event="KOSyncPullProgress", title=_("Pull progress from other devices"), rolling=true, paging=true, separator=true,},
-    page_jmp = { category="absolutenumber", event="GotoViewRel", min=-100, max=100, title=_("Go %1 pages"), rolling=true, paging=true,},
+    page_jmp = { category="absolutenumber", event="GotoViewRel", min=-100, max=100, title=_("Go %1 pages"), rolling=true, paging=true, separator=true,},
+    panel_zoom_toggle = { category="none", event="TogglePanelZoomSetting", title=_("Toggle panel zoom"), rolling=false, paging=true,},
```
Comment (yparitcher) [inline frontend/dispatcher.lua:107]: the separator will be shown after the item in the order of`dispatcher_menu_order` so there is no need to add it here, also there is no need to set `rolling=false`

```suggestion
    page_jmp = { category="absolutenumber", event="GotoViewRel", min=-100, max=100, title=_("Go %1 pages"), rolling=true, paging=true,},
    panel_zoom_toggle = { category="none", event="TogglePanelZoomSetting", title=_("Toggle panel zoom"), paging=true,},
```

File: frontend/dispatcher.lua [hunk 5]
```diff
@@ -206,6 +207,7 @@ local dispatcher_menu_order = {
     "back",
     "previous_location",
     "latest_bookmark",
+    "panel_zoom_toggle",
```
Comment (yparitcher) [inline frontend/dispatcher.lua:210]: this should probably go before or after
```lua
    "cycle_highlight_action",
    "cycle_highlight_style",
```

if you put it after then please remove the separator from `cycle_highlight_style` and add it to `panel_zoom_toggle`

**Review by yparitcher — CHANGES_REQUESTED** [review]

**Comment by Galunid** [comment]
To be fair, I don't really mind it in top menu, but I believe @Frenzie preferred it at the bottom.
I think I'll give up on that, since setting/getting default values is very problematic (no `hold_callback` and default value must be value (can't get it from function since `koptoptions` doesn't have `file` variable)). On first run we depend on the file extension. We also depend on extension when setting defaults.

**Comment by Frenzie** [comment]
I think it'd be easier in the bottom from a UX perspective but programmatically speaking it's most the same thing needing to support the same actions no matter where you put it.

File: frontend/dispatcher.lua [hunk 6]
```diff
@@ -100,10 +100,11 @@ local settingsList = {
...
     toggle_inverse_reading_order = { category="none", event="ToggleReadingOrder", title=_("Toggle page turn direction"), rolling=true, paging=true,},
     cycle_highlight_action = { category="none", event="CycleHighlightAction", title=_("Cycle highlight action"), rolling=true, paging=true,},
-    cycle_highlight_style = { category="none", event="CycleHighlightStyle", title=_("Cycle highlight style"), rolling=true, paging=true, separator=true,},
+    cycle_highlight_style = { category="none", event="CycleHighlightStyle", title=_("Cycle highlight style"), rolling=true, paging=true,},
     kosync_push_progress = { category="none", event="KOSyncPushProgress", title=_("Push progress from this device"), rolling=true, paging=true,},
     kosync_pull_progress = { category="none", event="KOSyncPullProgress", title=_("Pull progress from other devices"), rolling=true, paging=true, separator=true,},
     page_jmp = { category="absolutenumber", event="GotoViewRel", min=-100, max=100, title=_("Go %1 pages"), rolling=true, paging=true,},
+    panel_zoom_toggle = { category="none", event="TogglePanelZoomSetting", title=_("Toggle panel zoom"), rolling=false, paging=true, separator=true,},
```
Comment (yparitcher) [inline frontend/dispatcher.lua:107]:
```suggestion
    panel_zoom_toggle = { category="none", event="TogglePanelZoomSetting", title=_("Toggle panel zoom"), paging=true, separator=true,},
```

**Review by yparitcher — APPROVED** [review]

**Comment by Galunid** [comment]
>I think it'd be easier in the bottom from a UX perspective

@Frenzie 
I agree, but I don't think it's actually doable without a lot of hacky code, I think it's better left in the top menu, is that acceptable?

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 7]
```diff
@@ -107,6 +113,41 @@ local highlight_style = {
...
 
+local function getPanelZoomSupportedExt()
+    local default_supported_ext = {
+        cbz = true,
+        cbt = true,
+    }
+    return G_reader_settings:readSetting("panel_zoom_ext") or default_supported_ext
+end
+
+local function isDocumentComicOrManga(file)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:124]: Unless you plan later to detect if document is really a comic or manga, I would rename this to what it actually does :) 
`isPanelZoomSupported(file)`

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 8]
```diff
@@ -576,7 +617,37 @@ function ReaderHighlight:_resetHoldTimer(clear)
...
+        }
+        UIManager:show(imgviewer)
+    end
+    return true
+end
+
 function ReaderHighlight:onHold(arg, ges)
+    if self.panel_zoom_enabled then
+        self:onPanelZoom(arg, ges)
+        return false
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:648]: You're voluntarily letting the hold propagate?

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 9]
```diff
@@ -576,7 +617,37 @@ function ReaderHighlight:_resetHoldTimer(clear)
...
+            with_title_bar = false,
+            fullscreen = true,
+        }
+        UIManager:show(imgviewer)
+    end
+    return true
+end
+
 function ReaderHighlight:onHold(arg, ges)
+    if self.panel_zoom_enabled then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:646]: I would still wrap that into a `if self.document.info.has_pages then` to make it clear it works only on PDF and not on CRE.

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 10]
```diff
@@ -1298,6 +1369,12 @@ function ReaderHighlight:onReadSettings(config)
         disable_highlight = G_reader_settings:readSetting("highlight_disabled") or false
     end
     self.view.highlight.disabled = disable_highlight
+
+    -- panel zoom settings
+    self.panel_zoom_enabled = config:readSetting("panel_zoom_enabled")
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:1374]: I would wrap that too (loading & saving) into a `if self.document.info.has_pages then`.
(I can't update this setting as the menu is not displayed on CRE document, but if I were to toggle manually the saved setting in a metadata.epub.lua, things could break - so, I feel it's better to have explicite checks.)

**Review by poire-z — COMMENTED** [review]
I'm fine with letter that in the top menu.
If it happens it works nice and becore a top feature, we can then decide to move it into the bottom menu.

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 10 above)
Comment (Galunid) [reply frontend/apps/reader/modules/readerhighlight.lua:1374]: Good catch, since it's toggable via gesture

**Comment by Frenzie** [comment]
> I think it's better left in the top menu, is that acceptable?

Yes, it can always be changed later.

**Comment by Galunid** [comment]
Should be ready for the final review

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 11]
```diff
@@ -576,7 +617,37 @@ function ReaderHighlight:_resetHoldTimer(clear)
     end
 end
 
+function ReaderHighlight:onTogglePanelZoomSetting(arg, ges)
+    if not self.document.info.has_pages then return end
+    self.panel_zoom_enabled = not self.panel_zoom_enabled
+    self:onSaveSettings()
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:623]: I believe this onSaveSettings is unnecessary (it will be called when reader is closed and will save the updated self.panel_zoom_enabled ).

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 12]
```diff
@@ -576,7 +617,37 @@ function ReaderHighlight:_resetHoldTimer(clear)
...
+            with_title_bar = false,
+            fullscreen = true,
+        }
+        UIManager:show(imgviewer)
+    end
+    return true
+end
+
 function ReaderHighlight:onHold(arg, ges)
+    if self.panel_zoom_enabled and self.document.info.has_pages then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerhighlight.lua:647]: (nitpick: logical order, just for readability: first self.document.info.has_pages , and then self.panel_zoom_enabled  that makes sense only then.)

**Review by poire-z — APPROVED** [review]
lgtm (with these 2 nitpicks)

**Review by poire-z — APPROVED** [review]

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 13]
```diff
@@ -1298,6 +1368,14 @@ function ReaderHighlight:onReadSettings(config)
         disable_highlight = G_reader_settings:readSetting("highlight_disabled") or false
     end
     self.view.highlight.disabled = disable_highlight
+
+    -- panel zoom settings shouldn't work in EPUB
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerhighlight.lua:1372]: I'd phrase it slightly differently because if I look at my comics EPUB is a distinct download option! ;-D
```suggestion
    -- panel zoom settings isn't supported in EPUB
```

Screenshot_2020-09-24_15-15-05

**Review by Frenzie — APPROVED** [review]

**Comment by poire-z** [comment]
Just tested on the emulator with the only CBZ I have, it extracts things very fast !
But it seems there's some issue with the quality/scaling:
Original page in KOReader:
<kbd>image</kbd>
Zoomed panel in ImageViewer:
<kbd>image</kbd>
And in ImageViewer when I tap "Original size":
<kbd>image</kbd>

Dunno what's the reason (bad scaling by kopt/leptonica? bad scaling/dithering by our ImageWidget? better up-scaling by MuPDF? or my low DPI emulator?)

**Comment by Frenzie** [comment]
I thought the idea was to grab that part of the image at a higher resolution for optimal display, but that looks like it went wrong and grabbed lower res? I see the same effect actually.

Screenshot_2020-09-24_15-38-51
Screenshot_2020-09-24_15-39-14

**Comment by Frenzie** [comment]
But anyway, the panel detection works beautifully! ^_^

**Comment by Galunid** [comment]
@poire-z @Frenzie 
https://github.com/koreader/koreader/blob/15455b594d89810842aa0c1f6aa5d5431dfbcdb1/frontend/document/document.lua#L416-L421
Is responsible for cutting out the panel. I'm not really familiar with the whole blitbuffer/rendering stuff, so I don't know if it's possible to improve it. I guess there should be better way getting image, instad of using `Document:renderPage`, since that's most likely what's causing low resolution. Any suggestions are welcome.

**Comment by Frenzie** [comment]
It's not so much the renderPage as that you're passing a zoom of 1. What we want instead is something like this. :-)

https://github.com/koreader/koreader/blob/15455b594d89810842aa0c1f6aa5d5431dfbcdb1/frontend/document/koptinterface.lua#L285-L286

**Comment by Frenzie** [comment]
To make sure I get my implicit message across clearly, I think a `local zoom = math.min(canvas_size.w / rect.w, canvas_size.h / rect.h)` should do the trick but I'm on my laptop now so annoyingly I don't have an appropriate CBZ at hand to quickly test.

**Comment by Frenzie** [comment]
@Galunid Gave it a quick try, it works but you also need to adjust rect with the zoom factor after calculating it.

**Comment by Galunid** [comment]
```lua
function Document:getPagePart(pageno, rect, zoom, rotation)
    rect.x = rect.x * zoom
    rect.y = rect.y * zoom
    rect.w = rect.w * zoom
    rect.h = rect.h * zoom
    local tile = self:renderPage(pageno, rect, zoom, rotation, 1, 0)
    local target = Blitbuffer.new(rect.w, rect.h, self.render_color and self.color_bb_type or nil)
    target:blitFrom(tile.bb, 0, 0, rect.x, rect.y, rect.w, rect.h)
    return target
end
```

should work, but I'm not sure how to get canvas size from readerhighlight

**Comment by poire-z** [comment]
(I thought I could have mentionned the following - but only thing vaguely related is https://github.com/koreader/koreader/pull/6511#issuecomment-695798892.)

In there:
https://github.com/koreader/koreader/blob/c20ad8f5e9385dd18138a306b874c5b20302f299/frontend/apps/reader/modules/readerhighlight.lua#L685-L708
can't we tweak things a bit so that, when panel zoom is enabled, but no image is detected, just continue handling Hold the normal way and do dict lookup or text selection ?

Use case: I'm reading a PDF with a few illustations: with PanelZoom, long-press on an image shows it just like in EPUB, which is nice - but then dict doesn't work. I have to toggle panel zoom to do dict - and then toggle it again for the next image.
Would this be a problem with other contexts/use-cases ?

**Comment by Galunid** [comment]
@poire-z I was considering this when I first wrote this, the reason I opted not to do this is because in comics/manga (main application) it'd perform OCR, which is usually unwanted, and is pretty broken there. So yes, it'd be problem with comics/manga.

I don't know if we can detect text at given locations, since it's more of a cre/kopt stuff. I'm somewhat interested in being able to zoom PDF images, so it's possible I'll look a bit more into this in the near future.

**Comment by poire-z** [comment]
> I don't know if we can detect text at given locations

Before invoking OCR ? Dunno. It feels there should be a first step of detecting text boxes before invoking OCR, but all that might be handled as one step way far away :) 
May be the easier would be just another option in the `Panel zoom >` (where `[ ] Allow panel zoom` must feel a bit lonely :) : `[ ] Fallback to text selection when no image detected` - that I would enable, but comics/manga readers would not.

**Comment by Galunid** [comment]
I was thinking of something along those lines
